package com.fitness.gateway;
import java.text.ParseException;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ServerWebExchange;
import org.springframework.web.server.WebFilter;
import org.springframework.web.server.WebFilterChain;
import com.fitness.gateway.user.RegisterRequest;
import com.fitness.gateway.user.UserService;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.SignedJWT;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import reactor.core.publisher.Mono;

@Component
@Slf4j
@RequiredArgsConstructor
public class KeycloakUserSyncFilter implements WebFilter{
	
	private final  UserService  userService;
	
	
	@Override
	public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
        System.err.println("In filter method");
		String userId = exchange.getRequest().getHeaders().getFirst("X-User-ID");
		String token  = exchange.getRequest().getHeaders().getFirst("Authorization");
		
		RegisterRequest registerRequest = getUserDetails(token); 
		if(userId == null) {
			userId=registerRequest.getKeycloakId();
			System.err.println("user id : --"+userId);
			
		}
		if(userId!=null && token!=null) {
			String finalUserId=userId;
			return userService.validateUser(userId)
					.flatMap(exist->{
						if(!exist) {
							if(registerRequest!=null) {
								return userService.registerUser(registerRequest)
										.then(Mono.empty());
							}else {
								return Mono.<Void>empty();
							}
						}else {
							log.info("User already exists ,Skipping sync "); 
							return Mono.<Void>empty();

						}
					})
					.then(Mono.defer(() -> {
					    ServerHttpRequest mutatedRequest = exchange.getRequest().mutate()
					            .header("X-User-ID", finalUserId)
					            .build();

					    return chain.filter(exchange.mutate().request(mutatedRequest).build());
					}));
					
					
		}
		return null;
	}
	private RegisterRequest getUserDetails(String token) {
		System.err.println("in getUserDetailsof keycloak");
		try {
			String  tokenWithoutBearer = token.replace("Bearer", "").trim();
			SignedJWT signedJWT = SignedJWT.parse(tokenWithoutBearer);
			JWTClaimsSet claims = signedJWT.getJWTClaimsSet();
			RegisterRequest  request= new RegisterRequest();
			request.setEmail(claims.getStringClaim("email"));
			System.err.println("email: --"+claims.getStringClaim("email"));
			request.setKeycloakId(claims.getStringClaim("sub"));
			System.err.println("keycloakid : --"+claims.getStringClaim("sub"));
			request.setFirstName(claims.getStringClaim("given_name"));
			System.err.println("firstName: --"+claims.getStringClaim("given_name"));
			request.setLastName(claims.getStringClaim("family_name"));
			System.err.println("lastNsme : --"+claims.getStringClaim("family_name"));
			request.setPassword("pass@123123");
			
			return request;
			
		}catch(ParseException e){
			throw new RuntimeException(e);	
		}	
	}
}
